graph.go

```diff
@@ -4,193 +4,166 @@
+// edge is given from -> to, otherwise the edge is semantically
+// unordered.
 type Edge interface {
 	From() Node
 	To() Node
 }
 
-// A Graph implements the behavior of an undirected graph.
-//
-// All methods in Graph are implicitly undirected. Graph algorithms that care about directionality
-// will intelligently choose the DirectedGraph behavior if that interface is also implemented,
-// even if the function itself only takes in a Graph (or a super-interface of graph).
+// Graph is a generalized graph.
 type Graph interface {
-	// NodeExists returns true when node is currently in the graph.
-	NodeExists(Node) bool
-
-	// NodeList returns a list of all nodes in no particular order, useful for
-	// determining things like if a graph is fully connected. The caller is
-	// free to modify this list. Implementations should construct a new list
-	// and not return internal representation.
-	NodeList() []Node
-
-	// Neighbors returns all nodes connected by any edge to this node.
-	Neighbors(Node) []Node
-
-	// EdgeBetween returns an edge between node and neighbor such that
-	// From is one argument and To is the other. If no
-	// such edge exists, this function returns nil.
-	EdgeBetween(node, neighbor Node) Edge
-}
+	// Has returns whether the node exists within the graph.
+	Has(Node) bool
 
-// Directed graphs are characterized by having seperable Heads and Tails in their edges.
-// That is, if node1 goes to node2, that does not necessarily imply that node2 goes to node1.
-//
-// While it's possible for a directed graph to have fully reciprocal edges (i.e. the graph is
-// symmetric) -- it is not required to be. The graph is also required to implement Graph
-// because in many cases it can be useful to know all neighbors regardless of direction.
-type DirectedGraph interface {
-	Graph
-	// Successors gives the nodes connected by OUTBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Predecessors.
-	Successors(Node) []Node
-
-	// EdgeTo returns an edge between node and successor such that
-	// From returns node and To returns successor, if no
-	// such edge exists, this function returns nil.
-	EdgeTo(node, successor Node) Edge
-
-	// Predecessors gives the nodes connected by INBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Successors.
-	Predecessors(Node) []Node
-}
+	// Nodes returns all the nodes in the graph.
+	Nodes() []Node
 
-// Returns all undirected edges in the graph
-type EdgeLister interface {
-	EdgeList() []Edge
-}
+	// From returns all nodes that can be reached from
+	// the given node.
+	From(Node) []Node
 
-type EdgeListGraph interface {
-	Graph
-	EdgeLister
-}
+	// HasEdge returns whether an edge exists between
+	// nodes x and y without considering direction.
+	HasEdge(x, y Node) bool
 
-// Returns all directed edges in the graph.
-type DirectedEdgeLister interface {
-	DirectedEdgeList() []Edge
+	// Edge returns the edge between nodes u and v when
```

Comment:
This is not an explanation of From, it's a tie between Edge and From (which must agree for the API to work). It could repeat the definition used in From (reachability), and I considered that, but this emphasises the tie between them.

The alternative would be "Edge returns the edge from u to v if such an edge exists and nil otherwise. The node v must be directly reachable from u as defined by the From method."